src/creator.rs

```diff
@@ -77,74 +76,282 @@ impl<H: Hasher> Creator<H> {
         let parent_hashes: Vec<H::Hash> = parents.into_iter().flatten().collect();
 
         let new_preunit = PreUnit::new(self.node_ix, round, control_hash);
-        trace!(target: "AlephBFT-creator", "{:?} Created a new unit {:?} at round {:?}.", self.node_ix, new_preunit, self.current_round);
+        trace!(target: "AlephBFT-creator", "{:?} Created a new unit {:?} at round {:?}.", self.node_ix, new_preunit, round);
         self.new_units_tx
             .unbounded_send(NotificationOut::CreatedPreUnit(new_preunit, parent_hashes))
             .expect("Notification channel should be open");
 
-        self.current_round += 1;
-        self.init_round(self.current_round);
+        self.init_round(round + 1);
     }
 
     fn add_unit(&mut self, round: Round, pid: NodeIndex, hash: H::Hash) {
-        // units that are too old are of no interest to us
-        if round + 1 >= self.current_round {
-            self.init_round(round);
-            if self.candidates_by_round[round as usize][pid].is_none() {
-                // passing the check above means that we do not have any unit for the pair (round, pid) yet
-                self.candidates_by_round[round as usize][pid] = Some(hash);
-                self.n_candidates_by_round[round as usize] += NodeCount(1);
-            }
+        self.init_round(round);
+        if self.candidates_by_round[round as usize][pid].is_none() {
+            // passing the check above means that we do not have any unit for the pair (round, pid) yet
+            self.candidates_by_round[round as usize][pid] = Some(hash);
+            self.n_candidates_by_round[round as usize] += NodeCount(1);
         }
     }
 
-    fn check_ready(&self) -> bool {
-        if self.current_round == 0 {
-            return true;
-        }
-        if self.current_round > self.max_round {
-            warn!(target: "AlephBFT-creator", "{:?} Maximum round reached. Not creating another unit.", self.node_ix);
-            return false;
+    async fn wait_until_ready(&mut self, round: Round) {
+        let prev_round_index = match round.checked_sub(1) {
+            Some(prev_round) => prev_round as usize,
+            None => {
+                Delay::new((self.create_lag)(round.into())).await;
+                return;
```

Comment:
I don't like this `return` here. This does not allow to skip this delay. Instead one can move this piece of code to after the `loop` -- to line 128. Then you just remove the `Delay::new((self.create_lag)(round.into())).await;` and it should work fine.

Reply:
Done